### PR TITLE
Avoid error in target_market_share() when id_cols and values do not uniquely identify an observation

### PR DESCRIPTION
**Pull request by maurolepore**

### CODE REVIEW

**Comment by maurolepore** [comment]
``` r
library(tidyverse, warn.conflicts = FALSE)
library(r2dii.match)
#> Warning: package 'r2dii.match' was built under R version 4.0.2
devtools::load_all()
#> Loading r2dii.analysis
#> 
#> Attaching package: 'testthat'
#> The following object is masked from 'package:dplyr':
#> 
#>     matches
#> The following object is masked from 'package:purrr':
#> 
#>     is_null
#> The following object is masked from 'package:tidyr':
#> 
#>     matches
packageVersion("r2dii.analysis")
#> [1] '0.0.1.9000'

lbk_raw <- read_csv("~/Downloads/george/mock_lbk_to_reproduce.csv")
#> Warning: Missing column names filled in: 'X1' [1]
#> Parsed with column specification:
#> cols(
#>   .default = col_character(),
#>   X1 = col_double(),
#>   id_loan = col_double(),
#>   id_direct_loantaker = col_logical(),
#>   id_ultimate_parent = col_logical(),
#>   sector_classification_direct_loantaker = col_double(),
#>   fi_type = col_logical(),
#>   loan_size_outstanding_currency = col_logical(),
#>   loan_size_credit_limit_currency = col_logical(),
#>   loan_size_credit_limit = col_double(),
#>   loan_size_outstanding = col_double(),
#>   score = col_double()
#> )
#> See spec(...) for full column specifications.
# Select only the useful columns to avoid problems unrelated to this issue
lbk_cols <- c(
  "id_ultimate_parent",
  "name_ultimate_parent",
  "id_direct_loantaker",
  "name_direct_loantaker",
  "sector_classification_system",
  "sector_classification_direct_loantaker",
  "id_loan",
  "loan_size_outstanding"
)
lbk <- distinct(select(lbk_raw, all_of(lbk_cols)))

ald_raw <- read_csv("~/Downloads/george/ald_ff.csv")
#> Parsed with column specification:
#> cols(
#>   name_company = col_character(),
#>   sector = col_character(),
#>   technology = col_character(),
#>   year = col_double(),
#>   production = col_double(),
#>   production_unit = col_character(),
#>   emission_factor = col_logical(),
#>   ald_emission_factor_unit = col_logical(),
#>   plant_location = col_character(),
#>   is_ultimate_owner = col_logical(),
#>   ald_timestamp = col_character()
#> )
# Select only the useful columns to avoid problems unrelated to this issue
ald_cols <- c(
  "name_company",
  "sector",
  "technology",
  "year",
  "production",
  "is_ultimate_owner",
  "plant_location"
)
ald <- distinct(select(ald_raw, all_of(ald_cols)))

scen <- read_csv("~/Downloads/george/ff_sps_cps_sds_global.csv")
#> Parsed with column specification:
#> cols(
#>   scenario = col_character(),
#>   sector = col_character(),
#>   technology = col_character(),
#>   region = col_character(),
#>   year = col_double(),
#>   tmsr = col_double(),
#>   smsp = col_double(),
#>   scenario_source = col_character()
#> )



# The important part is this: target_market_share() no longer throws an error
matched <- prioritize(match_name(lbk, ald))
target_market_share(matched, ald, scen)
#> Joining, by = c("sector", "technology", "green_or_brown")
#> # A tibble: 310 x 7
#>    sector technology  year region scenario_source weighted_produc…
#>    <chr>  <chr>      <dbl> <chr>  <chr>           <chr>           
#>  1 oil a… gas         2020 global weo_2019        projected       
#>  2 oil a… gas         2020 global weo_2019        normalized_corp…
#>  3 oil a… gas         2020 global weo_2019        target_cps      
#>  4 oil a… gas         2020 global weo_2019        target_sds      
#>  5 oil a… gas         2020 global weo_2019        target_sps      
#>  6 oil a… oil         2020 global weo_2019        projected       
#>  7 oil a… oil         2020 global weo_2019        normalized_corp…
#>  8 oil a… oil         2020 global weo_2019        target_cps      
#>  9 oil a… oil         2020 global weo_2019        target_sds      
#> 10 oil a… oil         2020 global weo_2019        target_sps      
#> # … with 300 more rows, and 1 more variable: weighted_production_value <dbl>
```

<sup>Created on 2020-07-21 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by maurolepore** [comment]
@jdhoffa, I see some errors on CI but seem unrelated to our code. Whenever you are ready, please review and merge. Hopefully this will arrive on time to benefit @georgeharris2deg.

File: R/target_market_share.R [hunk 1]
```diff
@@ -158,14 +158,16 @@ target_market_share <- function(data,
       )
     ) %>%
     select(-.data$target_name, -.data$green_or_brown) %>%
-    tidyr::pivot_wider(
+    pivot_wider(
       names_from = .data$scenario,
       names_prefix = "weighted_production_target_",
-      values_from = .data$scenario_target_value
+      values_from = .data$scenario_target_value,
+      values_fn = list
```
Comment (jdhoffa) [inline R/target_market_share.R:165]: this is basically the only change, is that correct?

**Comment by jdhoffa** [comment]
I'm happy to approve this, but is it clear _why_ the `id_cols` and `value` are not uniquely identifying the observation? 
I worry @georgeharris2deg will end up with a list of two (or more) results for each entity, that are meant to be aggregated, and not be sure which to take...

File: R/target_market_share.R (hunk 1 above)
Comment (maurolepore) [reply R/target_market_share.R:165]: The changes are two: 
  * `values_fn = list` suppresses the warning by which `pivot_wider()`  tells the user that it had created a list column, automatically.
  * `unnest()` (next line) then transforms the list column into a regular column.

File: R/target_market_share.R [hunk 2]
```diff
@@ -158,14 +158,16 @@ target_market_share <- function(data,
...
     ) %>%
+    tidyr::unnest(starts_with("weighted_production_")) %>%
```
Comment (maurolepore) [inline R/target_market_share.R:167]: Here is `unnest()`

**Comment by maurolepore** [comment]
> I'm happy to approve this, but is it clear why the id_cols and value are not uniquely identifying the observation?

Not really. I don't understand enough -- so as to articulate it here -- why the demo datasets do not create such a list column but George's data does. I could give it another go today. But if you can it would be nice if you could fire a debugger near line 167 and compare what you get with `loanbook_demo` and `ald_demo` versus what you get with George's data.

**Comment by jdhoffa** [comment]
@maurolepore where can I find George's data? I'm happy to debug, but I don't have those datasets locally.

**Pull request merged**